### PR TITLE
feat: show cached images for instance creation [WD-14414]

### LINKED ISSUE #839: Show cached images during selection for instance creation
It would be nice to show which images have been downloaded previously during image selection for creating instances. This could be indicated in the image selection modal as shown below. @piperdeck perhaps this could be an extra column in the modal table, wdyt?

Screenshot from 2024-08-12 09-16-33

### PR DESCRIPTION
**Pull request by mas-who**
## Done

- Show cached images when creating instances

Closes #839 

## QA

1. Run the LXD-UI:
    - On the demo server via the link posted by @webteam-app below. This is only available for PRs created by collaborators of the repo. Ask @mas-who or @edlerd for access.
    - With a local copy of this branch, [build and run as described in the docs](../CONTRIBUTING.md#setting-up-for-development).
2. Perform the following QA steps:
    - Create an instance with a base image from any remote image server to download it for LXD
    - Then try create an instance again, make sure the previously downloaded image is available for selection in the image selection modal

## Screenshots
Screenshot from 2024-09-16 17-10-44

### CODE REVIEW

File: src/pages/images/ImageSelector.tsx [hunk 1]
```diff
@@ -403,6 +407,12 @@ const ImageSelector: FC<Props> = ({ onSelect, onClose }) => {
               ]}
               value={type ?? ""}
             />
+            <CheckboxInput
+              aria-label="Show cached images"
+              checked={showCached}
+              label="Show cached images"
+              onChange={() => setShowCached((prev) => !prev)}
+            />
```
Comment (edlerd) [inline src/pages/images/ImageSelector.tsx:415]: Do we need this, maybe it can be always on and doesn't need user control.

File: src/pages/images/ImageSelector.tsx [hunk 2]
```diff
@@ -272,6 +275,7 @@ const ImageSelector: FC<Props> = ({ onSelect, onClose }) => {
                 type="button"
                 dense
                 className="u-no-margin--bottom"
+                appearance={item.cached ? "positive" : "default"}
```
Comment (edlerd) [inline src/pages/images/ImageSelector.tsx:278]: Should we also have the green one for local images?

File: src/util/images.tsx [hunk 3]
```diff
@@ -48,9 +48,17 @@ export const localLxdToRemoteImage = (image: LxdImage): RemoteImage => {
     arch: image.architecture === "x86_64" ? "amd64" : image.architecture,
     os: image.properties?.os ?? "",
     created_at: new Date(image.uploaded_at).getTime(),
-    release: image.properties?.release ?? "",
-    server: LOCAL_IMAGE,
+    release:
+      image.properties?.os === "ubuntu"
+        ? (image.properties.version ?? "")
+        : (image.properties?.release ?? ""),
```
Comment (edlerd) [inline src/util/images.tsx:54]:
```suggestion
```
Why is this needed?
There is similar special handling of Ubuntu in the image selector component. The goal is to have it read nicely "Ubuntu xx.xx codename" as people are used to that ordering.

File: src/util/images.tsx [hunk 4]
```diff
@@ -48,9 +48,17 @@ export const localLxdToRemoteImage = (image: LxdImage): RemoteImage => {
...
+    server: image.cached ? image.update_source?.server : LOCAL_IMAGE,
     type: image.type,
+    cached: image.cached,
+    variant:
+      image.properties?.os === "ubuntu"
+        ? image.properties?.release
+        : image.properties?.variant,
```
Comment (edlerd) [inline src/util/images.tsx:61]:
```suggestion
```
Seems not needed to me. Can it be removed?

File: src/pages/images/ImageSelector.tsx (hunk 1 above)
Comment (mas-who) [reply src/pages/images/ImageSelector.tsx:415]: This was added after the discussion with @piperdeck . I think it could be useful to have the option to filter out cached images?

File: src/util/images.tsx (hunk 3 above)
Comment (mas-who) [reply src/util/images.tsx:54]: This is added so that we can populate the "Release" column within the modal. The ternary condition is needed because for an `LxdImage` returned from the server, the data is not organised exactly the same as a `RemoteImage`. The field `image.properties.version` in this case for a cached `ubuntu` image aligns with the "Release" column content populated with the remote image data.

Comment (edlerd) [reply src/util/images.tsx:54]: Ah I see, this is handling the cached entries and transforming them into the same format as the ones we fetch from image servers.

Comment (edlerd) [reply src/util/images.tsx:54]: Still suspicious if we should have the special handling of Ubuntu here and in the modal.

Comment (edlerd) [reply src/util/images.tsx:54]: Suggesting to make this funciton more straight forward:
```
export const localLxdToRemoteImage = (image: LxdImage): RemoteImage => {
  return {
    aliases: image.update_source?.alias ?? image.aliases?.[0]?.name ?? "",
    fingerprint: image.fingerprint,
    arch: image.architecture === "x86_64" ? "amd64" : image.architecture,
    os: image.properties?.os ?? "",
    created_at: new Date(image.uploaded_at).getTime(),
    release: image.properties?.release ?? "",
    release_title: image.properties?.version ?? "",
    server: LOCAL_IMAGE,
    type: image.type,
  };
};
```
And keep the special handling of ubuntu in the ImageSelector.
If we pass the os here in this function through a converter, that ensures the first character is uppercase, it all should work well.

Comment (mas-who) [reply src/util/images.tsx:54]: yeah agreed, will implement that. I would keep the `cached` field though since we want to differentiate later on whether a local image is a cached remote image or uploaded image.

File: src/util/images.tsx (hunk 4 above)
Comment (mas-who) [reply src/util/images.tsx:61]: closing this since the discussion is happening on the other comment.

File: src/pages/images/ImageSelector.tsx (hunk 2 above)
Comment (edlerd) [reply src/pages/images/ImageSelector.tsx:278]: We can instead sort the cached images to the top of the list and keep the current behaviour of the "postive" and default button appearance.

File: src/pages/images/ImageSelector.tsx (hunk 1 above)
Comment (mas-who) [reply src/pages/images/ImageSelector.tsx]: this is now removed

File: src/pages/images/ImageSelector.tsx (hunk 2 above)
Comment (mas-who) [reply src/pages/images/ImageSelector.tsx:276]: implemented the sorting :+1:

File: src/pages/images/ImageSelector.tsx [hunk 5]
```diff
@@ -210,19 +210,20 @@ const ImageSelector: FC<Props> = ({ onSelect, onClose }) => {
...
         if (item.server === minimalServer) {
-          return "Ubuntu Minimal";
+          source = "Ubuntu Minimal";
         }
         if (item.server === imagesLxdServer) {
-          return "LXD Images";
+          source = "LXD Images";
         }
-        return "Custom";
+        return source + (item.cached ? " (cached)" : "");
```
Comment (edlerd) [inline src/pages/images/ImageSelector.tsx:226]: I think we can better present this as a small muted text to not draw too much attention through the two lines with brackets

image

**Review by edlerd — COMMENTED** [review]
QA and code LGTM

One idea to improve the design.

File: src/pages/images/ImageSelector.tsx (hunk 5 above)
Comment (mas-who) [reply src/pages/images/ImageSelector.tsx:226]: great idea, done! :+1:

File: src/pages/images/ImageSelector.tsx [hunk 6]
```diff
@@ -210,19 +210,25 @@ const ImageSelector: FC<Props> = ({ onSelect, onClose }) => {
...
+          source = "LXD Images";
         }
-        return "Custom";
+        return (
+          <>
+            {source}{" "}
+            {item.cached ? <span className="u-text--muted">cached</span> : ""}
+          </>
+        );
       };
```
Comment (edlerd) [inline src/pages/images/ImageSelector.tsx:232]: Shall we keep this function to return the string as before. Then append the cached string based on the column below in the column? Seems a bit simpler to me.

**Review by edlerd — APPROVED** [review]
QA and code LGTM, just one comment on a possible change, but can also merge as is.

File: src/pages/images/ImageSelector.tsx (hunk 6 above)
Comment (mas-who) [reply src/pages/images/ImageSelector.tsx]: done :+1:

File: src/pages/images/ImageSelector.tsx [hunk 7]
```diff
@@ -260,7 +261,12 @@ const ImageSelector: FC<Props> = ({ onSelect, onClose }) => {
             onClick: selectImage,
           },
           {
-            content: getSource(),
+            content: (
+              <>
+                {getSource()}{" "}
+                {item.cached && <span className="u-text--muted">cached</span>}
```
Comment (edlerd) [inline src/pages/images/ImageSelector.tsx:267]: The space should be inside the conditional.

**Review by edlerd — APPROVED** [review]
All good, thanks for adding the cached image visibility.

File: src/pages/images/ImageSelector.tsx (hunk 7 above)
Comment (mas-who) [reply src/pages/images/ImageSelector.tsx]: thanks! didn't pick that up as "cached" were all on a new line for me

**Review by edlerd — APPROVED** [review]

**Comment by mas-who** [comment]
@edlerd had a discussion with @piperdeck and made the following updates:

1. Added a checkbox filter to "Show only cached images" if checked. The thinking here is that if the user for some reason have many cached images, and they want to sort those images for some reason (instead of search), then having the ability to sort cached images would be good. For remote images, we expect the user would  always want to do a search and filter really. I'm not entirely sure about this one TBH, but we agreed to implement it since it's quick and see what you think.

2. The "Select" button for local images (ones that we create in LXD or uploaded) also now have positive appearance.

**Comment by piperdeck** [comment]
Please change the checkbox to "Only show cached images" instead of "Show only cached images". Then it's good to go!
